File: Web/src/api/backtest.ts
import { api } from "./client";

export interface BacktestSummary {
  total_trades: number;
  win_rate: number;
  avg_return: number;
  median_return: number;
  big_win_rate: number;
  big_loss_rate: number;
  avg_hold_days: number;
  sharpe?: number;
  max_drawdown_pct?: number;
  calmar?: number;
  cagr_pct?: number;
  initial_capital?: number;
  total_profit?: number;
  final_capital?: number;
  signal_count?: number;
  skipped_count?: number;
  max_concurrent?: number;
}

export interface MonthlyReturn {
  year: number;
  month: number;
  return_pct: number;
}

export interface EquityPoint {
  date: string;
  nav: number;
}

export interface BacktestMetrics {
  sharpe: number;
  max_drawdown_pct: number;
  calmar: number;
  cagr_pct: number;
  monthly: MonthlyReturn[];
  equity_curve: EquityPoint[];
  initial_capital?: number;
  total_profit?: number;
  final_capital?: number;
}

export interface BacktestTask {
  id: string;
  name: string | null;
  strategy_name: string;
  strategy_params: Record<string, unknown>;
  start_date: string;
  end_date: string;
  take_profit: number;
  stop_loss: number;
  max_hold: number;
  split_tp: number | null;
  initial_capital: number;
  position_pct: number;
  max_concurrent: number;
  t_plus_1: boolean;
  status: "pending" | "running" | "done" | "error" | "cancelled";
  progress: number;
  total: number;
  error: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  elapsed_seconds: number | null;
  summary: BacktestSummary | null;
  trade_count: number;
}

export interface BacktestTrade {
  code: string;
  name: string;
  signal_date: string;
  market: string;
  tier: string;
  score: number;
  breakout_pct: number;
  is_limit_up: boolean;
  vol_ratio: number;
  macd: number;
  dif: number;
  pullback_pct: number;
  ma_spread_pct: number;
  days_since_test: number;
  close_to_ma30: number;
  close_to_low60: number;
  body_ratio: number;
  day_change_pct: number;
  bull_ma_count: number;
  buy_price: number;
  buy_date: string;
  sell_price: number;
  sell_date: string;
  sell_reason: string;
  return_pct: number;
  max_up_pct: number;
  max_dn_pct: number;
  hold_days: number;
  quantity: number;
  buy_amount: number;
  sell_amount: number;
  profit_amount: number;
}

export interface LedgerRow {
  date: string;
  action: "buy" | "sell";
  code: string;
  name: string;
  signal_date: string;
  buy_date?: string | null;
  price: number;
  quantity: number;
  amount: number;
  profit_amount: number | null;
  sell_reason: string | null;
}

export interface LedgerPage {
  rows: LedgerRow[];
  total: number;
  page: number;
  page_size: number;
  initial_capital: number;
  total_profit: number;
  final_capital: number;
}

export interface BacktestRequest {
  name?: string | null;
  strategy: string;
  params: Record<string, unknown>;
  start_date: string;
  end_date: string;
  take_profit: number;
  stop_loss: number;
  max_hold: number;
  split_tp?: number | null;
  max_codes?: number | null;
  num_workers?: number | null;
  engine?: "legacy" | "vectorbt";
  initial_capital?: number;
  position_pct?: number;
  max_concurrent?: number;
  t_plus_1?: boolean;
}

export interface TradesPage {
  rows: BacktestTrade[];
  total: number;
  page: number;
  page_size: number;
}

export async function createBacktest(req: BacktestRequest): Promise<{ task_id: string }> {
  const { data } = await api.post<{ task_id: string }>("/backtest", req);
  return data;
}

export async function getBacktest(taskId: string): Promise<BacktestTask> {
  const { data } = await api.get<BacktestTask>(`/backtest/${taskId}`);
  return data;
}

export async function listBacktestHistory(limit = 50): Promise<BacktestTask[]> {
  const { data } = await api.get<{ tasks: BacktestTask[] }>("/backtest/history", {
    params: { limit },
  });
  return data.tasks;
}

export async function listBacktestTrades(
  taskId: string,
  opts: { page?: number; pageSize?: number; sortBy?: string; desc?: boolean } = {}
): Promise<TradesPage> {
  const { data } = await api.get<TradesPage>(`/backtest/${taskId}/trades`, {
    params: {
      page: opts.page ?? 1,
      page_size: opts.pageSize ?? 50,
      sort_by: opts.sortBy ?? "score",
      desc: opts.desc ?? true,
    },
  });
  return data;
}

export async function deleteBacktest(taskId: string): Promise<void> {
  await api.delete(`/backtest/${taskId}`);
}

export async function getBacktestMetrics(taskId: string): Promise<BacktestMetrics> {
  const { data } = await api.get<BacktestMetrics>(`/backtest/${taskId}/metrics`);
  return data;
}

export async function listBacktestLedger(
  taskId: string,
  opts: { page?: number; pageSize?: number } = {}
): Promise<LedgerPage> {
  const { data } = await api.get<LedgerPage>(`/backtest/${taskId}/ledger`, {
    params: {
      page: opts.page ?? 1,
      page_size: opts.pageSize ?? 100,
    },
  });
  return data;
}

export type WsMessage =
  | { type: "snapshot"; task_id: string; status: string; done: number; total: number; trade_count: number; summary: BacktestSummary | null; error: string | null; elapsed_seconds: number | null }
  | { type: "progress"; task_id: string; done: number; total: number; trade_count: number; elapsed_seconds: number }
  | { type: "done"; task_id: string; summary: BacktestSummary; trade_count: number; elapsed_seconds: number }
  | { type: "error"; task_id: string; error: string };

export function subscribeBacktest(
  taskId: string,
  onMessage: (m: WsMessage) => void,
  onClose?: () => void
): () => void {
  // Vite 代理 ws：开发模式下用 ws://localhost:5173/ws/...，proxy 自动转发
  const proto = window.location.protocol === "https:" ? "wss:" : "ws:";
  const url = `${proto}//${window.location.host}/ws/backtest/${taskId}`;
  const ws = new WebSocket(url);

  ws.addEventListener("message", (ev) => {
    try {
      const msg = JSON.parse(ev.data) as WsMessage;
      onMessage(msg);
    } catch (e) {
      console.warn("invalid ws message", e);
    }
  });
  ws.addEventListener("close", () => onClose?.());

  return () => ws.close();
}

File: backend/app/schemas/backtest.py
"""回测 schemas。"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ..core.time_utils import UtcDateTime


class BacktestRequest(BaseModel):
    name: str | None = Field(default=None, description="任务名（可空，便于历史中区分）")
    strategy: str = "breakout_washout"
    params: dict[str, Any] = Field(default_factory=dict)
    start_date: str = Field(description="YYYY-MM-DD")
    end_date: str = Field(description="YYYY-MM-DD")
    take_profit: float = Field(default=0.20, gt=0)
    stop_loss: float = Field(default=0.07, gt=0)
    max_hold: int = Field(default=20, ge=1, le=120)
    split_tp: float | None = Field(default=None, description="分批止盈，如 0.05；None 表示不分批")
    max_codes: int | None = Field(default=None, description="调试用：仅扫描前 N 只股票")
    num_workers: int | None = Field(default=8, ge=1, le=32, description="并行进程数，默认 8")
    engine: str = Field(default="legacy", description="模拟引擎：'legacy'（精确）或 'vectorbt'（实验性，sl/tp 基于 close）")
    initial_capital: float = Field(default=1_000_000.0, gt=0, description="初始资金（元）")
    position_pct: float = Field(default=1.0, gt=0, le=1.0, description="单笔仓位占初始资金比例，1.0=每笔用满")
    max_concurrent: int = Field(default=1, ge=1, le=20, description="最大同时持仓只数，1=串行全仓")
    t_plus_1: bool = Field(default=True, description="A 股 T+1：买入当日不可卖出")


class BacktestSummary(BaseModel):
    total_trades: int
    win_rate: float
    avg_return: float
    median_return: float
    big_win_rate: float
    big_loss_rate: float
    avg_hold_days: float
    sharpe: float = 0.0
    max_drawdown_pct: float = 0.0
    calmar: float = 0.0
    cagr_pct: float = 0.0
    initial_capital: float = 0.0
    total_profit: float = 0.0
    final_capital: float = 0.0
    signal_count: int = 0
    skipped_count: int = 0
    max_concurrent: int = 0


class MonthlyReturn(BaseModel):
    year: int
    month: int
    return_pct: float


class EquityPoint(BaseModel):
    date: str
    nav: float


class BacktestMetrics(BaseModel):
    sharpe: float
    max_drawdown_pct: float
    calmar: float
    cagr_pct: float
    monthly: list[MonthlyReturn]
    equity_curve: list[EquityPoint]
    initial_capital: float = 0.0
    total_profit: float = 0.0
    final_capital: float = 0.0


class BacktestTaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str | None
    strategy_name: str
    strategy_params: dict
    start_date: str
    end_date: str
    take_profit: float
    stop_loss: float
    max_hold: int
    split_tp: float | None
    initial_capital: float = 1_000_000.0
    position_pct: float = 1.0
    max_concurrent: int = 1
    t_plus_1: bool = True
    status: str
    progress: int
    total: int
    error: str | None
    created_at: UtcDateTime
    started_at: UtcDateTime | None
    finished_at: UtcDateTime | None
    elapsed_seconds: float | None
    summary: dict | None
    trade_count: int


class BacktestCreateResponse(BaseModel):
    task_id: str
    status: str = "pending"


class BacktestListResponse(BaseModel):
    tasks: list[BacktestTaskOut]


class BacktestTradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    code: str
    name: str = ""
    signal_date: str
    market: str
    tier: str
    score: float
    breakout_pct: float
    is_limit_up: bool
    vol_ratio: float
    macd: float
    dif: float
    pullback_pct: float
    ma_spread_pct: float
    days_since_test: int
    close_to_ma30: float
    close_to_low60: float
    body_ratio: float
    day_change_pct: float
    bull_ma_count: int
    buy_price: float
    buy_date: str
    sell_price: float
    sell_date: str
    sell_reason: str
    return_pct: float
    max_up_pct: float
    max_dn_pct: float
    hold_days: int
    quantity: int = 0
    buy_amount: float = 0.0
    sell_amount: float = 0.0
    profit_amount: float = 0.0


class LedgerRow(BaseModel):
    date: str
    action: str  # buy | sell
    code: str
    name: str = ""
    signal_date: str
    buy_date: str | None = None  # 卖出行携带对应买入日，便于 K 线双标记
    price: float
    quantity: int
    amount: float
    profit_amount: float | None = None
    sell_reason: str | None = None


class LedgerPage(BaseModel):
    rows: list[LedgerRow]
    total: int
    page: int
    page_size: int
    initial_capital: float = 0.0
    total_profit: float = 0.0
    final_capital: float = 0.0


class TradesPage(BaseModel):
    rows: list[BacktestTradeOut]
    total: int
    page: int
    page_size: int
